plugins/legion-hid2/fu-legion-hid2-device.c

```diff
@@ -0,0 +1,522 @@
+					guint16 tag,
+					GError **error)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autoptr(FuChunk) chk = NULL;
+		g_autoptr(GByteArray) req = fu_struct_legion_iap_tlv_new();
+		g_autoptr(GByteArray) res = fu_struct_legion_iap_tlv_new();
+
+		fu_struct_legion_iap_tlv_set_tag(req, tag);
+
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+
+		if (!fu_memcpy_safe(req->data,
+				    req->len,
+				    FU_STRUCT_LEGION_IAP_TLV_OFFSET_VALUE,
+				    fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk),
+				    0x0,
+				    fu_chunk_get_data_sz(chk),
+				    error))
+			return FALSE;
+
+		fu_struct_legion_iap_tlv_set_length(req, fu_chunk_get_data_sz(chk));
+
+		if (!fu_legion_hid2_device_tlv(self, req, res, error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_wait_for_complete_cb(FuDevice *device, gpointer user_data, GError **error)
+{
+	g_autoptr(GByteArray) cmd = fu_struct_legion_iap_tlv_new();
+	g_autoptr(GByteArray) result = fu_struct_legion_iap_tlv_new();
+	const guint8 *value;
+
+	fu_struct_legion_iap_tlv_set_tag(cmd, FU_LEGION_IAP_HOST_TAG_IAP_CARRY);
+
+	if (!fu_legion_hid2_device_tlv(FU_LEGION_HID2_DEVICE(device), cmd, result, error)) {
+		g_prefix_error(error, "failed to verify code: ");
+		return FALSE;
+	}
+	value = fu_struct_legion_iap_tlv_get_value(result, NULL);
+	if (value[1] < 100) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_BUSY,
+			    "device is %d percent done",
+			    value[1]);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_write_data(FuLegionHid2Device *self,
+				 FuFirmware *firmware,
+				 FuProgress *progress,
+				 GError **error)
+{
+	guint32 offset;
+	gssize length;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) data = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	offset = fu_legion_hid2_firmware_get_data_offset(firmware);
+	length = fu_legion_hid2_firmware_get_data_size(firmware);
+	data = g_bytes_new_from_bytes(fw, offset, length);
+	chunks = fu_chunk_array_new_from_bytes(data, 0, FU_STRUCT_LEGION_IAP_TLV_SIZE_VALUE);
+	if (!fu_legion_hid2_device_write_data_chunks(self,
+						     chunks,
+						     progress,
+						     FU_LEGION_IAP_HOST_TAG_IAP_DATA,
+						     error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_write_sig(FuLegionHid2Device *self,
+				FuFirmware *firmware,
+				FuProgress *progress,
+				GError **error)
+{
+	guint32 offset;
+	gssize length;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) data = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
```

Comment:
can you use a stream instead?

Reply:
I started to convert it and it wasn't working, and then I remember what was wrong.  fu_chunk_array_new_from_stream() returns too much data.  By doing the `GBytes` dance It can be bounded to only pull exactly what is needed.
Any suggestion on a way to bound the chunk array for a stream?  Or maybe `fu_chunk_array_new_from_stream()` needs an argument for length?

Reply:
mmm, I started that and when I was changing I realized I should instead make partial image streams in the firmware parser, will go this direction instead.

Reply:
Yup, partial image stream is exactly what you want. Then it's zero copy!